### PR TITLE
Modifying Rating model and added tests.

### LINKED ISSUE #36: Create Model stubs for developer reference
We should define some model stubs in our tasties_app so we can have a shared baseline for everyone to develop the models from.

- [x] Define model classes in models.py
- [x] Add fields according to ERD

### LINKED ISSUE #41: Implementing "Rating" class
### Description
After the last issue where we declared our models' classes, now it's time to implement them and create migrations and tests for them.
This issue will handle the "Rating" class.

### Tasks
- [ ] Define the new class "Rating" and its relevant fields and functionalities in models.py.
- [ ] Create a migrations file in the "migrations" directory and test the data.
- [ ] Create tests file using "pytest" and write efficient tests.

### PR DESCRIPTION
**Pull request by ohadrevach**
Defined Rating model and added update function for it.
  Also added tests for adding,removing and updating rating.
Close #41 
Close #36

### CODE REVIEW

**Review by GalOved — CHANGES_REQUESTED** [review]
Looking good to me good work.
If you look in the test_rating file you can see that you need too import a specific models. after that flake8 will pass.
And I would appreciate it if you shared with me why you chose to make static methods? Is there a reason for this that will help us in the future?
I don't think it's anything wrong, I just want to understand if I might also need to do it.

**Review by GalOved — APPROVED** [review]
Looking good to me

**Review by efraimvis — COMMENTED** [review]
Please link this PR to an issue. Also, in case [this](https://github.com/redhat-beyond/tasties/pull/46) PR will get merged, I think you should consider removing 0001_initial.py from this PR.

**Review by WaseemAbuSobe — APPROVED** [review]

**Review by natifridman — COMMENTED** [review]
This PR is for the Rating model, do we need other models here as well?

File: tasties_app/models.py [hunk 1]
```diff
@@ -1,3 +1,89 @@
...
+
+class Rating(models.Model):
+    author_id = models.ForeignKey(User, on_delete=models.CASCADE)
+    recipe_id = models.ForeignKey(Recipe, on_delete=models.CASCADE)
+    rating = models.IntegerField(
+        validators=[MinValueValidator(1), MaxValueValidator(5)]
+    )
+
+    @staticmethod
+    def add_rating(author_id, recipe_id, rating):
```
Comment (natifridman) [inline tasties_app/models.py:71]: Why do you need this method if you don't use it?

File: tasties_app/models.py [hunk 2]
```diff
@@ -1,3 +1,89 @@
...
+        """
+        users = User.objects.filter(username=author_id.username)
+        if len(users) > 0:
+            raise ValueError("Already voted")
+            return
+        rating = Rating(author_id, recipe_id, rating)
+        rating.save()
+        return rating
+
+    def edit_rating(self, new_rate):
```
Comment (natifridman) [inline tasties_app/models.py:84]: Maybe instead of `edit`, `update` will explain better what this method doing

File: tasties_app/models.py [hunk 3]
```diff
@@ -1,3 +1,89 @@
...
+        self.rating = new_rate
+
+    @staticmethod
+    def remove_rating(rating):
```
Comment (natifridman) [inline tasties_app/models.py:88]: Why do you need delete method if you can just do `rating.delete()`

File: tasties_app/tests/test_rating.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+class TestRatingModel:
+    @pytest.mark.django_db
+    def test_add_rating(self, generate_data):
+        rating_set = Rating.objects.all()
+        rating = rating_set[0]
+        assert rating.rating == 3
+
+    @pytest.mark.django_db
+    def test_delete(self, generate_data):
```
Comment (natifridman) [inline tasties_app/tests/test_rating.py:38]: Please give the test a meaningful name, something like this
```suggestion
    def test_delete_rating(self, rating):
```

File: tasties_app/tests/test_rating.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        rating_set = Rating.objects.all()
+        rating = rating_set[0]
+        rating.remove_rating(rating)
+        assert len(rating_set) == 0
+
+    @pytest.mark.django_db
+    def test_edit(self, generate_data):
```
Comment (natifridman) [inline tasties_app/tests/test_rating.py:45]: Please give the test a meaningful name, something like this
```suggestion
    def test_edit_rating(self, rating):
```

**Review by GilShechter — APPROVED** [review]
LGTM, nice work

**Comment by ohadrevach** [comment]
> This PR is for the Rating model, do we need other models here as well?

Yes, because Rating uses other models so it was needed to define them.

**Review by efraimvis — DISMISSED** [review]

**Comment by liornoy** [comment]
The commit description is a bit misleading: "Added functions for remov, adding and editing rating,"
Where you just added a function for updating the rating.
Also if you added all the other model classes, why not add the corresponding comments to everyone?
like you did for the rating class:
```
    """
    This Class represent Rating object
    Fields:
        author_id (User): User object.
        recipe_id (Recipe): Recipe object.
        rating (int): rating for the recipe, 1-5
    """
```

**Comment by liornoy** [comment]
Also, this PR resolves issue #36, please add it to the PR description.

**Comment by ohadrevach** [comment]
> The commit description is a bit misleading: "Added functions for remov, adding and editing rating," Where you just added a function for updating the rating. Also if you added all the other model classes, why not add the corresponding comments to everyone? like you did for the rating class:
> 
> ```
>     """
>     This Class represent Rating object
>     Fields:
>         author_id (User): User object.
>         recipe_id (Recipe): Recipe object.
>         rating (int): rating for the recipe, 1-5
>     """
> ```

Updated the description, I didn't write to anyone because everyone did this comments to the model he worked on.

**Comment by liornoy** [comment]
Please reword the `Modifying Rating model` commit message, as you didn't add remove and add functions.

**Comment by ohadrevach** [comment]
> Please reword the `Modifying Rating model` commit message, as you didn't add remove and add functions.

Done

**Comment by liornoy** [comment]
You changed the PR title. I asked for the commit message of the "Modifying Rating model" commit.

**Comment by ohadrevach** [comment]
> You changed the PR title. I asked for the commit message of the "Modifying Rating model" commit.

changed it back

File: tasties_app/models.py [hunk 6]
```diff
@@ -1,3 +1,85 @@
...
+    ML = "ML"
+    UNIT_CHOICES = (
+        (WHOLE, "Whole"),
+        (FLOZ, "Fluid Ounce"),
+        (TSP, "Teaspoon"),
+        (OZ, "Ounce"),
+        (CUP, "Cup"),
+        (GRAM, "Gram"),
+        (ML, "Milliliter"),
+    )
```
Comment (natifridman) [inline tasties_app/models.py:44]: Why not use enum for this?

**Review by natifridman — COMMENTED** [review]
I would also add some tests here, what if I create a rating of 6 for example? or without an author?

**Comment by liornoy** [comment]
> > You changed the PR title. I asked for the commit message of the "Modifying Rating model" commit.
> 
> changed it back

You still haven't updated that commit message.
In this PR you have 2 commits:
1. Populated models.py with model stubs, updated Pipfiles
2. Modifying Rating model 

The commit message of commit 2 is:
```
Added functions for remov, adding and editing rating,
Added tests for all of the above.
```

This message is not relevant as you didn't add "remove" and "adding" functions.
Therefore it's a misleading commit message.
Please reword this commit message using the git rebase reword command.

**Comment by ohadrevach** [comment]
> > > You changed the PR title. I asked for the commit message of the "Modifying Rating model" commit.
> > 
> > 
> > changed it back
> 
> You still haven't updated that commit message. In this PR you have 2 commits:
> 
> 1. Populated models.py with model stubs, updated Pipfiles
> 2. Modifying Rating model
> 
> The commit message of commit 2 is:
> 
> ```
> Added functions for remov, adding and editing rating,
> Added tests for all of the above.
> ```
> 
> This message is not relevant as you didn't add "remove" and "adding" functions. Therefore it's a misleading commit message. Please reword this commit message using the git rebase reword command.

Understood, fixed it as requested

File: tasties_app/tests/test_rating.py [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    def test_delete_rating(self, rating):
+        rating.delete()
+        assert len(Rating.objects.all()) == 0
+
+    @pytest.mark.django_db
+    def test_edit_rating(self, rating):
+        rating_set = Rating.objects.all()
+        rating = rating_set[0]
+        rating.update_rating(4)
+        assert rating.rating == 4
```
Comment (natifridman) [inline tasties_app/tests/test_rating.py:46]: Is this checking the DB? I see that you override `rating` and check the same object

Comment (liornoy) [reply tasties_app/tests/test_rating.py:42]: I think what it does is test the update function and assert that it updates the rating field of the rating object.

Comment (ohadrevach) [reply tasties_app/tests/test_rating.py:42]: I did what Lior said

**Review by GilShechter — APPROVED** [review]

**Review by efraimvis — APPROVED** [review]

File: tasties_app/tests/test_rating.py (hunk 7 above)
Comment (natifridman) [reply tasties_app/tests/test_rating.py:42]: Do we want to check that the update is saved?

File: tasties_app/tests/test_rating.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+    @pytest.mark.django_db
+    def test_add_bad_rating(self, rating):
+        rating_set = Rating.objects.all()
+        rating = rating_set[0]
+        with pytest.raises(ValueError):
+            rating.update_rating(6)
+
+    @pytest.mark.django_db
+    def test_add_bad_rating_aid(self, rating):
```
Comment (natifridman) [inline tasties_app/tests/test_rating.py:52]: It's unclear from the name, what is the difference between this test and the other test with the same name?

File: tasties_app/tests/test_rating.py (hunk 7 above)
Comment (ohadrevach) [reply tasties_app/tests/test_rating.py:42]: Yes, i'm checking that this method is working fine

File: tasties_app/tests/test_rating.py (hunk 8 above)
Comment (ohadrevach) [reply tasties_app/tests/test_rating.py:61]: > It's unclear from the name, what is the difference between this test and the other test with the same name?

Added docstring for the functions

Comment (natifridman) [reply tasties_app/tests/test_rating.py:61]: We usually try to make the test name self-explained. I think you can change the name to something like 
```suggestion
    def test_invalid_rating_author_id(self, rating):
```

**Review by natifridman — COMMENTED** [review]
Please don't force push during the review, it makes it hard to understand what you changed from last time.

File: tasties_app/models.py [hunk 9]
```diff
@@ -1,3 +1,82 @@
...
+
+    author_id = models.ForeignKey(User, on_delete=models.CASCADE)
+    recipe_id = models.ForeignKey(Recipe, on_delete=models.CASCADE)
+    rating = models.IntegerField(
+        validators=[MinValueValidator(1), MaxValueValidator(5)]
+    )
+
+    def update_rating(self, new_rate):
+        if type(new_rate) != int or new_rate > 5 or new_rate < 1:
+            raise ValueError("Invalid value")
```
Comment (liornoy) [inline tasties_app/models.py:75]: This should be under the function's comment.

**Review by GilShechter — APPROVED** [review]

**Review by efraimvis — APPROVED** [review]

**Review by GalOved — APPROVED** [review]
LGTM

**Review by natifridman — APPROVED** [review]
LGTM, nice work!

**Pull request merged**